File: sdk/python-can-gsusb/examples/04_custom_timing.py
import can

from common import get_demo_custom_timings, get_parser
from gsusb import GsUsbBus


def main() -> None:
    parser = get_parser("Single-channel custom timing CAN FD example")
    parser.add_argument("--gsusb-channel", type=int, default=0, help="Logical gs_usb channel (default: 0)")
    args = parser.parse_args()

    nominal, data = get_demo_custom_timings()

    print(f"Opening {args.selector} Channel {args.gsusb_channel}...")
    print("Applying custom timing: nominal=(2,0,31,8,8), data=(1,0,7,2,2)")
    bus = GsUsbBus(
        channel=args.selector,
        gsusb_channel=args.gsusb_channel,
        fd=True,
        nominal_timing=nominal,
        data_timing=data,
        termination_enabled=not args.no_termination,
    )
    try:
        msg = can.Message(
            arbitration_id=0x7F0,
            is_extended_id=False,
            is_fd=True,
            bitrate_switch=True,
            data=[0x11] + [0x00] * 11,
        )
        print("Sending custom timing frame...")
        bus.send(msg)
        print("Listening...")
        while True:
            rx = bus.recv(1.0)
            if rx is not None:
                print(f"Received: {rx}")
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        bus.shutdown()


if __name__ == "__main__":
    main()

File: sdk/gsusb-rs/examples/04_custom_timing.rs
#[path = "common/mod.rs"]
mod common;

use common::{get_demo_custom_timings, make_bus_options, make_payload, parse_common_options, print_message};
use gsusb_rs::{CanMessage, GsUsbBus, Result};
use std::env;
use std::time::Duration;

fn main() -> Result<()> {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let options = parse_common_options(&args, "Single-channel custom timing CAN FD example")?;
    let (nominal, data) = get_demo_custom_timings();
    let tx_bus = GsUsbBus::new(make_bus_options(&options, options.tx_channel, true, Some(nominal), Some(data)))?;
    let rx_bus = GsUsbBus::new(make_bus_options(&options, options.rx_channel, true, Some(nominal), Some(data)))?;

    let message = CanMessage {
        arbitration_id: 0x7F0,
        is_fd: true,
        bitrate_switch: true,
        data: make_payload(&[0x11], 12, 0x00),
        ..CanMessage::default()
    };
    print_message("Sending", &message);
    tx_bus.send(&message)?;
    let received = rx_bus.recv(Duration::from_secs(2)).ok_or_else(|| gsusb_rs::GsUsbError("No custom timing frame received".to_string()))?;
    print_message("Received", &received);
    tx_bus.shutdown();
    rx_bus.shutdown();
    Ok(())
}